pandas/core/categorical.py

```diff
@@ -2243,6 +2219,12 @@ def _get_codes_for_values(values, categories):
     """
 
     from pandas.core.algorithms import _get_data_algo, _hashtables
+    if is_categorical_dtype(values.dtype):
+        codes = (values.cat.codes if isinstance(values, ABCSeries)
+                 else values.codes)
```

Comment:
There _must_ be a cleaner way... but I couldn't find it.

Reply:
Or the question could be: why does this get passed values that are already categorical and don't need a factorize but a recode ?

Reply:
meaning that the check could also be done before (which I think is how it was before?)

something like

```
if dtype.categories is None:
   codes, categories = ....
elif is_categorical_dtype(values):
    .... handle this case
else:
    codes = _get_codes_for_values(values, dtype.categories)
```

I personally would find that easier to follow the logic (and this does not necessarily mean you don't need the `values.cat.codes` vs `values.codes` ..., so my comment went a bit sideways :-))